mqtt-protocol/src/jvmNativeMain/kotlin/it/unibo/pulvreakt/mqtt/MqttProtocol.kt

```diff
@@ -70,12 +69,21 @@ actual class MqttProtocol actual constructor(
         either {
             val topic = registeredTopics[Pair(from, to)]
             logger.debug { "Writing message $message to topic $topic" }
+
             ensureNotNull(topic) { ProtocolError.EntityNotRegistered(to) }
-            val mqttMessage = MqttMessage(message).apply { qos = 2 }
-            async(coroutineDispatcher) {
-                Either.catch { mqttClient.publish(topic, mqttMessage).waitForCompletion() }
-                    .mapLeft { ProtocolError.ProtocolException(it) }
-            }.await().bind()
+
+            Either.catch {
+                client.publish(
+                    retain = true,
+                    qos = if (qos <= 0) Qos.AT_MOST_ONCE else if (qos >= 2) Qos.EXACTLY_ONCE else Qos.AT_LEAST_ONCE,
```

Comment:
Move out the `QoS` assignment.
Usually, I prefer using a `when` instead of a multi-if.